File: src/socket/socket_12_ping_icmp_protocol.c
#include <fcntl.h>
#include <sys/socket.h>
#include <resolv.h>
#include <netdb.h>
#include <netinet/in.h>
#include <netinet/ip_icmp.h>
#include <stdbool.h>
#include <unistd.h>
#include <string.h>

struct packet {
    struct icmphdr hdr;
    char msg[64 - sizeof(struct icmphdr)];
};

// checksum - standard 1s complement checksum
unsigned short checksum(void *b, int len) {
    unsigned short *buf = b;
    unsigned int sum = 0;
    unsigned short result;
    for (sum = 0; len > 1; len -= 2)
        sum += *buf++;
    if (len == 1)
        sum += *(unsigned char *) buf;
    sum = (sum >> 16) + (sum & 0xFFFF);
    sum += (sum >> 16);
    result = ~sum;
    return result;
}

bool ping(char *adress) {
    struct sockaddr_in r_addr;
    struct sockaddr_in addr_ping, *addr;
    struct hostent *hname = gethostbyname(adress);
    bzero(&addr_ping, sizeof(addr_ping));
    addr_ping.sin_family = hname->h_addrtype;
    addr_ping.sin_port = 0;
    addr_ping.sin_addr.s_addr = *(long *) hname->h_addr;
    addr = &addr_ping;

    int sd = socket(AF_INET, SOCK_DGRAM, IPPROTO_ICMP);
//    int sd = socket(PF_INET, SOCK_RAW, IPPROTO_ICMP);
    if (sd < 0) {
        perror("socket");
        return false;
    }
    if (fcntl(sd, F_SETFL, O_NONBLOCK) != 0) {
        perror("fcntl");
        return false;
    }
    const int ttl_val = 64;
    if (setsockopt(sd, SOL_IP, IP_TTL, &ttl_val, sizeof(ttl_val)) != 0) {
        perror("setsockopt");
        return false;
    }

//    int cnt = 1;
    for (int loop = 0; loop < 10; loop++) {
        int len = sizeof(r_addr);
        printf("before recvfrom\n");
        struct packet pckt;
        if (recvfrom(sd, &pckt, sizeof(pckt), 0, (struct sockaddr *) &r_addr, &len) > 0) {
            return true;
        }
        printf("after recvfrom\n");

        bzero(&pckt, sizeof(pckt));
        pckt.hdr.type = ICMP_ECHO;
//        int i = 0;
        for (int i = 0; i < sizeof(pckt.msg) - 1; i++)
            pckt.msg[i] = i + '0';
//        pckt.msg[i] = 0;
//        pckt.hdr.un.echo.sequence = cnt++;
        pckt.hdr.checksum = checksum(&pckt, sizeof(pckt));
        printf("checksum = %d\n", pckt.hdr.checksum);
        if (sendto(sd, &pckt, sizeof(pckt), 0, (struct sockaddr *) addr, sizeof(*addr)) <= 0)
            perror("sendto");
        usleep(300 * 1000);
    }
    return false;
}

int main() {
    if (ping("www.baidu.com") == true) {
        printf("Ping is OK. \n");
    } else {
        printf("Ping is not OK. \n");
    }
    return 0;
}

File: src/socket/socket_12_ping_icmp_protocol.rs
use crate::{gethostbyname, inet_ntoa, syscall, SOCKADDR_IN_LEN};

#[allow(clippy::ptr_as_ptr, clippy::cast_ptr_alignment)]
fn nslookup(hostname: &str) -> libc::sockaddr_in {
    let hostname = std::ffi::CString::new(hostname).unwrap();
    let hostname = hostname.as_ptr().cast();
    let hostent = unsafe { gethostbyname(hostname) };
    if hostent.is_null() {
        panic!("Invalid hostname or DNS lookup lookup failed");
    }
    let hostent = unsafe { *hostent };
    let remote_addr = unsafe { *(*hostent.h_addr_list as *mut libc::in_addr) };
    syscall!(printf(
        "PING %s (%s) 64 bytes of data\n\0".as_ptr().cast(),
        hostname,
        inet_ntoa(remote_addr),
    ));
    libc::sockaddr_in {
        sin_family: libc::AF_INET as libc::sa_family_t,
        sin_port: 0,
        sin_addr: remote_addr,
        sin_zero: unsafe { std::mem::zeroed() },
    }
}

const ICMP_ECHO: u8 = 8;

#[repr(C)]
struct Packet {
    type_: u8,
    code: u8,
    checksum: u16,
    union_padding: u32,
    msg: [u8; PACKET_LEN - 8],
}

const PACKET_LEN: usize = 64;

#[test]
fn main() {
    let remote_addr = nslookup("baidu.com");

    // SOCK_RAW need sudo/root permission
    // let sockfd = libc::socket(libc::AF_INET, libc::SOCK_RAW, (*libc::getprotobyname("ICMP\0".as_ptr().cast())).p_proto);
    let sockfd = syscall!(socket(libc::AF_INET, libc::SOCK_DGRAM, libc::IPPROTO_ICMP));
    syscall!(fcntl(sockfd, libc::F_SETFL, libc::O_NONBLOCK));
    syscall!(setsockopt(
        sockfd,
        libc::SOL_IP,
        libc::IP_TTL,
        (&64 as *const i32).cast(),
        std::mem::size_of::<i32>() as u32
    ));

    for _ in 0..10 {
        let mut packet: Packet = unsafe { std::mem::zeroed() };
        let mut addr = remote_addr;
        let mut addrlen = SOCKADDR_IN_LEN;
        println!("before recv");
        let recvfrom_ret = unsafe {
            libc::recvfrom(
                sockfd,
                (&mut packet as *mut Packet).cast(),
                PACKET_LEN,
                0,
                (&mut addr as *mut libc::sockaddr_in).cast(),
                &mut addrlen,
            )
        };
        if recvfrom_ret > 0 {
            println!("ping success");
            std::process::exit(libc::EXIT_SUCCESS);
        }
        println!("after recv");

        packet = unsafe { std::mem::zeroed() };
        packet.type_ = ICMP_ECHO;
        for i in 0..packet.msg.len() - 1 {
            packet.msg[i] = i as u8 + b'0';
        }
        println!("packet.msg.len() = {}", packet.msg.len());
        println!("packet.msg = {:?}", packet.msg);
        // 先写死，在 linux_command_rewritten_in_rust 的 repo 去实现 pnet::util::checksum 算法

        // packet.checksum = 3772;
        packet.checksum = 62527; // 用 3772 或 62527 都行

        // packet.hdr.checksum = pnet::util::checksum(&packet.msg, 0);

        syscall!(sendto(
            sockfd,
            (&packet as *const Packet).cast(),
            PACKET_LEN,
            0,
            (&remote_addr as *const libc::sockaddr_in).cast::<libc::sockaddr>(),
            SOCKADDR_IN_LEN,
        ));

        syscall!(usleep(300 * 1000));
    }
    eprintln!("ping failed!");
}
